api/clients/v2/validator/retrieval_worker.go

```diff
@@ -0,0 +1,606 @@
+	}
+
+	targetDownloadCount := uint32(math.Ceil(float64(minimumChunkCount) * config.DownloadPessimism))
+	targetVerifiedCount := uint32(math.Ceil(float64(minimumChunkCount) * config.VerificationPessimism))
+
+	worker := &retrievalWorker{
+		ctx:                       ctx,
+		downloadAndVerifyCtx:      downloadAndVerifyCtx,
+		cancel:                    cancel,
+		logger:                    logger,
+		config:                    config,
+		connectionPool:            connectionPool,
+		computePool:               computePool,
+		encodingParams:            encodingParams,
+		assignments:               assignments,
+		quorumID:                  quorumID,
+		blobKey:                   blobKey,
+		verifier:                  verifier,
+		blobCommitments:           blobCommitments,
+		downloadStartedChan:       make(chan *downloadStarted, len(assignments)),
+		downloadCompletedChan:     make(chan *downloadCompleted, len(assignments)),
+		verificationCompletedChan: make(chan *verificationCompleted, len(assignments)),
+		decodeResponseChan:        make(chan *decodeCompleted, 1),
+		probe:                     probe,
+		minimumChunkCount:         minimumChunkCount,
+		downloadsInProgressQueue:  linkedlistqueue.New(),
+		downloadedChunksQueue:     linkedlistqueue.New(),
+		verifiedChunksQueue:       linkedlistqueue.New(),
+		downloadOrder:             downloadOrder,
+		totalChunkCount:           totalChunkCount,
+		chunkStatusMap:            chunkStatusMap,
+		chunkStatusCounts:         chunkStatusCounts,
+		targetDownloadCount:       targetDownloadCount,
+		targetVerifiedCount:       targetVerifiedCount,
+		validatorGRPCManager:      validatorGRPCManager,
+		chunkDeserializer:         chunkDeserializer,
+		blobDecoder:               blobDecoder,
+	}
+
+	return worker, nil
+}
+
+// updateChunkStatus updates the status of a chunk from a given operator. It also updates the
+// counts of the various chunk statuses.
+func (w *retrievalWorker) updateChunkStatus(operatorID core.OperatorID, status chunkStatus) {
+	operatorChunks := w.assignments[operatorID].NumChunks
+
+	oldStatus, ok := w.chunkStatusMap[operatorID]
+	if ok {
+		w.chunkStatusCounts[oldStatus] -= int(operatorChunks)
+	}
+	w.chunkStatusMap[operatorID] = status
+	w.chunkStatusCounts[status] += int(operatorChunks)
+}
+
+// getChunkStatus returns the status of a chunk from a given operator.
+func (w *retrievalWorker) getChunkStatus(operatorID core.OperatorID) chunkStatus {
+	return w.chunkStatusMap[operatorID]
+}
+
+// getStatusCount returns the number of chunks with one of the given statuses.
+func (w *retrievalWorker) getStatusCount(statuses ...chunkStatus) uint32 {
+	total := 0
+	for _, status := range statuses {
+		if count, ok := w.chunkStatusCounts[status]; ok {
+			total += count
+		}
+	}
+	return uint32(total)
+}
+
+// downloadBlobFromValidators downloads the blob from the validators.
+func (w *retrievalWorker) downloadBlobFromValidators() ([]byte, error) {
```

Comment:
nit: I think a better name for this method might be `retrieveBlobFromValidators`

- this is a `retrievalWorker`, so it makes sense that the verb of its main entrypoint be `retrieve`
- this struct makes a distinction between downloading and verifying. By including the word "download", and not the word "verify", in this method name, it gives the impression that this method is *only* related to the download step, and that verification is handled elsewhere. But in reality, this method does the downloading, *and* the verifying

Reply:
renamed